src/util/ConfigManager/ConfigManager.cpp

```diff
@@ -0,0 +1,447 @@
+    const {
+  return configurationOptions_;
+}
+
+// ____________________________________________________________________________
+const ConfigOption& ConfigManager::getConfigurationOptionByNestedKeys(
+    const VectorOfKeysForJson& keys) const {
+  // If there is an entry at the described location, this should point to the
+  // index number of the configuration option in `configurationOptions_`.
+  const nlohmann::json::json_pointer ptr{createJsonPointer(keys)};
+
+  if (keyToConfigurationOptionIndex_.contains(ptr)) {
+    return configurationOptions_.at(
+        keyToConfigurationOptionIndex_.at(ptr).get<size_t>());
+  } else {
+    throw NoConfigOptionFoundException(vectorOfKeysForJsonToString(keys),
+                                       printConfigurationDoc(true));
+  }
+}
+
+// ____________________________________________________________________________
+nlohmann::json ConfigManager::parseShortHand(
+    const std::string& shortHandString) {
+  // I use ANTLR expressions to parse the short hand.
+  antlr4::ANTLRInputStream input(shortHandString);
+  ConfigShorthandLexer lexer(&input);
+  antlr4::CommonTokenStream tokens(&lexer);
+  ConfigShorthandParser parser(&tokens);
+
+  // The default in ANTLR is to log all errors to the console and to continue
+  // the parsing. We need to turn parse errors into exceptions instead to
+  // propagate them to the user.
+  ThrowingErrorListener errorListener{};
+  parser.removeErrorListeners();
+  parser.addErrorListener(&errorListener);
+  lexer.removeErrorListeners();
+  lexer.addErrorListener(&errorListener);
+
+  // Get the top node. That is, the node of the first grammar rule.
+  ConfigShorthandParser::ShortHandStringContext* shortHandStringContext{
+      parser.shortHandString()};
+
+  // Walk through the parser tree and build the json equivalent out of the short
+  // hand.
+  return ToJsonConfigShorthandVisitor{}.visitShortHandString(
+      shortHandStringContext);
+}
+
+// ____________________________________________________________________________
+void ConfigManager::parseConfig(const nlohmann::json& j) {
+  // Anything else but a literal json object is not something, we want.
+  if (!j.is_object()) {
+    throw ConfigManagerParseConfigNotJsonObjectLiteralException(j);
+  }
+
+  /*
+  According to the documentation, when iterating over the entries of a
+  `nlohmann::json` object using `items()`, there can be error/problems, IF the
+  life time of the object, on which one called `items()` on, doesn't exceeds the
+  life time of the iteration. (See the documentation for more information.)
+
+  Because we iterate over the 'flattend' verions of two `nlohmann::json`
+  objects, I gave them const references here.
+  */
+  const auto& jFlattend = j.flatten();
+  const auto& keyToConfigurationOptionIndexFlattend =
+      keyToConfigurationOptionIndex_.flatten();
+
+  // We can skip the following check, if `j` is empty.
+  if (!j.empty()) {
```

Comment:
> Is this `if` necessary? If `j` is empty, then the flattened version is also empty, and thus the `for ` loop in line 237 iterates over nothing. So I think it can be dropped.

Funny thing: I thought the exact same thing, but **nope**. The flattend version is **not** empty. (It's been a while, but, if I remember correctly, it resulted in a `{"":null}`.)

Because an empty `nlohmann::json`-object shouldn't cause an exception, when, for example, we only had `ConfigOption` with default values, I added that `if`-clause. Otherwise, we always get an exception about `""` not pointing to an existing `ConfigOption`.

Reply:
Thanks for pointing this out, maybe add the comment about this for people who are not that much into `nlohmann::json`:
```
// Note: even if the JSON object is empty, its flattened version contains a single dummy entry, so this check is necessary.
```

Reply:
Good idea. Done, see [cb22262](https://github.com/ad-freiburg/qlever/pull/977/commits/cb22262688144a54f000d2001ead5511886b65d5).